### PR TITLE
Jira: OAuth

### LINKED ISSUE #51: Jira OAuth
Right now Jira is using basic auth, but OAuth is the preferred option (and that way we don't have to save anyone's credentials).

### PR DESCRIPTION
**Pull request by epaulet**
fixes #51

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,31 @@
...
+  def cors_preflight_check
+    if request.method == 'OPTIONS'
+      if ALLOWED_ORIGINS.include?(request.headers['HTTP_ORIGIN'])
+        headers['Access-Control-Allow-Origin'] = request.headers['HTTP_ORIGIN']
+        headers['Access-Control-Allow-Credentials'] = 'true'
+        headers['Access-Control-Allow-Methods'] = 'POST, GET, PUT, DELETE, OPTIONS'
+        headers['Access-Control-Allow-Headers'] = 'accept,content-type'
+      end
+
+      render :text => '', :content_type => 'text/plain'
```
Comment (shosti) [inline app/controllers/application_controller.rb]: Is this a standard thing, to have an empty response?

Comment (epaulet) [inline app/controllers/application_controller.rb]: Yeah. Although I think I'm technically missing a 'Content-Length' field. I'll double check that.

**Comment by epaulet** [comment]
@shosti anything else :)

**Comment by shosti** [comment]
lol yeah taking another look...

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,5 +1,31 @@
 class ApplicationController < ActionController::Base
+  ALLOWED_ORIGINS = ENV.fetch('ALLOWED_ORIGINS').split(',').freeze
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   # protect_from_forgery with: :exception
+
+  after_action :cors_set_access_control_headers
```
Comment (shosti) [inline app/controllers/application_controller.rb]: out of curiosity, why is this `after_action` instead of before?

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,9 @@
 Rails.application.routes.draw do
+  match '/*all', to: 'application#cors_preflight_check', via: [:options]
```
Comment (shosti) [inline config/routes.rb]: What does this route match? Aren't routes regex? (In which case this would match `/////all` and such.)

File: app/controllers/oauth_controller.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  def access_token
+    settings = session[:oauth].to_hash.with_indifferent_access
+    session[:oauth] = nil
+
+    consumer = create_consumer(settings)
+    request_token = OAuth::RequestToken.new(consumer, settings[:oauth_token], settings[:oauth_token_secret])
+    access_token = request_token.get_access_token(oauth_verifier: params[:oauth_verifier])
+
+    # Can't return consumer secret because it's too long for the query string
```
Comment (shosti) [inline app/controllers/oauth_controller.rb]: What do you mean by this? Why would you return the secret anyways?

File: app/controllers/oauth_controller.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    # Return the authentication information needed for future OAuth Access
+    redirect_to "#{FRONTEND_URL}/settings/integrations?#{returned_params.to_query}"
+  end
+
+  private
+
+  def create_consumer(settings)
+    OAuth::Consumer.new(
+      settings[:consumer_key],
+      OpenSSL::PKey::RSA.new(settings[:consumer_secret]),
```
Comment (shosti) [inline app/controllers/oauth_controller.rb]: Oy, the OpenSSL API is horrible, does `PKey` mean public or private?

**Comment by shosti** [comment]
@epaulet can you describe the overall authentication flow (and where the main app fits into it)? It's kind of hard to follow from the main app...

File: app/controllers/oauth_controller.rb (hunk 4 above)
Comment (epaulet) [inline app/controllers/oauth_controller.rb]: I would return it so they don't have to type it in the settings again. You can't save your settings without an access token and secret first in Rainforest.

File: app/controllers/oauth_controller.rb (hunk 5 above)
Comment (epaulet) [inline app/controllers/oauth_controller.rb]: It creates a pair.

File: config/routes.rb (hunk 3 above)
Comment (epaulet) [inline config/routes.rb]: Yeah but only for OPTIONS requests.

Comment (shosti) [inline config/routes.rb]: OIC nvm (forgot that the routing syntax is weird).

**Comment by epaulet** [comment]
Sure.

1. regenwald sends OAuth settings over to rainforest-integrations (consumer key, consumer secret, oauth endpoints, etc.) to `/oauth/request-token`
2. rainforest-integrations take that information and retrieves the request token from the authorization server and tells the server the oauth callback should go to `/oauth/access-token`. It stores the oauth information in the user's session hash (this is recommended in the documentation I've seen) and responds to regenwald by telling it the url where the user can authenticate.
3. regenwald opens a new tab/window with the authentication url, allowing the user to authenticate.
4. The authorization server authenticates the user and creates an oauth verifier parameter, and sends that information to the callback that was setup in step 2. (`/oauth/access-token`)
5. rainforest-integrations receives the oauth parameter, and with the information stored in the user's session, is able to finally retrieve the access token. It then sends that back to regenwald.
6. regenwald receives the information and is able to ask the user for any other settings required.

Why so much going on in this app? Because all of the tokens have to be retrieved on the backend because some authentication servers aren't set up for CORS (I don't think OAuth requires it). And since this doesn't require any sort of DB, I figured I would have rainforest-integrations handle the authentication rather than our backend.

**Comment by epaulet** [comment]
http://sharing.rainforestqa.com/3k0b2A3S3S1W

**Comment by shosti** [comment]
But what credentials are stored eventually, and where?

**Comment by shosti** [comment]
Do you have a link to the auth docs for JIRA? (Sorry to be a PITA with this, there's just a fairly high risk if this gets screwed up so I want to be sure...)

**Comment by shosti** [comment]
My big question is whether this flow should happen against rainforest-integrations or against the main app (because that's where credentials will eventually be stored, right?). Any particular reason it has to be against the integrations app?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (shosti) [inline app/controllers/application_controller.rb]: Pretty sure Rack/Rails should handle that (although never hurts to double-check). Also I think `render :text` implies `Content-Type: text/plain`.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (epaulet) [inline app/controllers/application_controller.rb]: because browser requests will refuse to use the response if the correct access isn't granted in the headers of a cors request. But if we end up moving the authentication to Rainforest, then this `after_action` will only need to happen on the one endpoint that the frontend uses.

**Comment by epaulet** [comment]
@shosti ripped the controller out of here and cleaned up the unnecessary headers.

**Comment by shosti** [comment]
:ship:
